src/UndisclosedPassword.php

```
@@ -0,0 +1,177 @@
+            return false;
+        } catch (\Exception $exception) {
+            $this->error(self::UNKNOWN_ERROR);
+            return false;
+        }
+        if ($isPwnd) {
+            $this->error(self::PASSWORD_BREACHED);
+            return false;
+        }
+        return true;
+    }
+
+    private function isPwnedPassword($password)
+    {
+        $sha1Hash = $this->hashPassword($password);
+        $rangeHash = $this->getRangeHash($sha1Hash);
+        $hashList = $this->retrieveHashList($rangeHash);
+        return $this->hashInResponse($sha1Hash, $hashList);
+    }
+
+    /**
+     * We use a SHA1 hashed password for checking it against
+     * the breached data set of HIBP.
+     *
+     * @param string $password
+     * @return string
+     */
+    private function hashPassword($password)
+    {
+        $hashedPassword = \sha1($password);
+        return strtoupper($hashedPassword);
+    }
+
+    /**
+     * Creates a hash range that will be send to HIBP API
+     * applying K-Anonymity
+     *
+     * @param string $passwordHash
+     * @return string
+     * @see https://www.troyhunt.com/enhancing-pwned-passwords-privacy-by-exclusively-supporting-anonymity/
+     */
+    private function getRangeHash($passwordHash)
+    {
+        $range = substr($passwordHash, self::HIBP_RANGE_BASE, self::HIBP_RANGE_LENGTH);
+        return $range;
+    }
+
+    /**
+     * Making a connection to the HIBP API to retrieve a
+     * list of hashes that all have the same range as we
+     * provided.
+     *
+     * @param string $passwordRange
+     * @return string
+     * @throws Exception\RuntimeException
+     */
+    private function retrieveHashList($passwordRange)
+    {
+        $request = $this->httpRequest->createRequest('GET', '/range/' . $passwordRange);
+
+        try {
+            $response = $this->httpClient->sendRequest($request);
+        } catch (ClientExceptionInterface $connectException) {
```

You are catching/re-throwing and then catching in the caller again here: if we can recognise precise `ClientExceptionInterface` types, let's react to them in the caller, and otherwise let exceptions bubble up, since otherwise misconfiguration scenarios become extremely hard to inspect.